source/transfer_list.rst

```diff
@@ -364,15 +364,57 @@ Inputs:
    to `tl.checksum` (`new_tl_base + 0x4`).
 
 
-Entry type ranges
------------------
-
-The content of the data section is determined by the tag id. The tag id space contains two ranges:
-
- #. Standard range, and
- #. Non-standard range
-
-The `tag_id` ranges are described in :numref:`tab_tag_id_ranges`.
+Entry type allocation
+---------------------
+
+Tag IDs must be allocated in this specification before use. A new tag ID can be
+allocated by submitting a pull request to this repository that adds a
+description of the respective TE data layout to this specification. Tag IDs do
+not have to be allocated in order. Submitters are encouraged to try to group
+tag IDs together in logical clusters at 16 or 256-aligned boundaries (e.g. all
+tags related to a particular chipset or to a particular firmware project could
+use adjacent tag numbers), but there are no predefined ranges and no
+reservations of tag ranges for specific use.
+
+Tags are expected to have a simple layout (representable by a C structure) and
+each tag should only represent data for a single logical concept. Data for
+multiple distinct concepts should be split across different tags, even if
+they're always expected to appear together on the first platform adding the tag
+(to encourage reusability in different situations). Alternatively, complex data
+may be represented in a different kind of well-established handoff data
+structure (e.g. FDT, HOB) that is inserted into the TL as a single TE. The same
```

Comment:
For FDT can you please keep the link to the spec? Also you've replaced 'self-describing' with 'well-established' which is not the same thing. Can you put that back?

HOB would be very confusing since it includes pointers and is tied to use of EFI...so I suggest just dropping that and saying (e.g. FDT)

Reply:
I don't think this specification should be passing judgement on which other handoff formats are "good" or "bad". Different projects may be tied to different legacy structures that regardless of what disadvantages they may have still need to be used for certain reasons. The point here is to say "if you want to describe stuff in the transfer list directly, follow these rules -- otherwise, you can also wrap your existing legacy strucutre in a TE and keep using that for some data". I am not very familiar with HOB and am not trying to champion it specifically but I think this guidance should be solution-agnostic, so it would be good to list more than one example to illustrate that. Since there's already a TE defined for HOBs they are clearly already allowed in this role?

I don't really understand what "self-describing" is supposed to mean in this context, or why it is needed. What makes the FDT self-describing in a matter that other structures are not? (Are HOBs "self-describing"?) An FDT is still a binary (i.e. not all-text) structure that I need a parser to read, after all. Do you just mean "a key-value store with ASCII keys"? Requiring that here seems unnecessarily narrow to me.

Reply:
I agree it would be good to have spec links but I think this is a minor comment. Otherwise looks good to me. We know of at least 1 partner who wants to use HOBs in the TL. I tend to agree the self-describing comment doesn't add much value; I think this just refers to the fact that the structure contains a header with a signature. In that sense, the TL (and all specified TEs) are self-describing, even if individual TEs are not.

Reply:
We may have forgotten why this was there.

The first thing that came up for "self-describing data structure" just now was "A message that contains data as well as the metadata that describes the format and the meaning (i.e., the syntax and the semantics) of that data. For example, Extensible Markup Language (XML) is a self-describing message format that consists of tag/value pairs."

Clearly a HOB Is not self-describing. You must have the format to decode the data. In fact even that might be hard. For example, each HOB entry has a GUID which indicates its type, but these can be private and unknown.

The goal here is, as I understand it, to avoid people creating a new binary structure that cannot be read by other projects, except for pre-existing formats, i.e. HOB and ACPI. The reason for mentioning FDT is that it can be read without needing a secret spec or code, etc.

We already have HOB list in the spec though, as a tag type:

https://github.com/FirmwareHandoff/firmware_handoff/blob/main/source/transfer_list.rst#hob-list-entry-layout-xferlist_hob_l

I am OK with submitting this without the link to the FDT spec if that helps. I can do a fix-up PR later.

So, please, can we have something like:

Alternatively, complex data may be represented in a self-describing data format such as the FDT [DT]_.

Reply:
> The goal here is, as I understand it, to avoid people creating a new binary structure that cannot be read by other projects, except for pre-existing formats, i.e. HOB and ACPI

Hmm, I thought the goal was simply to avoid people creating a new binary structure when a suitable one already exists. Although HOB/ACPI are not self-describing, I don't think that means very much in this context because all binary structures become self-describing when wrapped in a TE specified in this spec. So HOB is a reasonable choice for some implementations and worthy of mention IMO.

Reply:
> The reason for mentioning FDT is that it can be read without needing a secret spec or code, etc.

Sorry, but this is just completely untrue. The FDT is a binary structure and if you don't have the right structure definitions for decoding the FDT header, node structure, property structure, reserve list, string table and all that gunk, it is just as much of an opaque blob to you as a HOB. That is my fundamental objection to making FDT stand out like it was something special and "better" here, because it isn't. It's a hierarchical key-value store with ASCII keys and integer-array or ASCII values. HOB lists are a key-value store with GUID keys and memory buffer values. (The TL itself, for comparison, is a key-value store with integer keys and C structure values.) Those are not the same thing, but I don't see how the difference matters enough for this context to the point where we need to tell people to prefer one over the other. If they want to be interoperable, they should be using TEs directly anyway, and if they want to stick to whatever other legacy structures their stack is already used to, they should be free do that too.

An XML document is different because it includes a link to a schema doc that describes the meaning of the fields (which an FDT has not), so I assume that's why wherever you pulled that quote from calls it "self-describing" (not that I think that this property was very useful for our purposes here either).

Reply:
I won't respond to the first paragraph other than to say FDT is considerably less opaque than a binary blob.

Dan, the point of this text was to avoid people creating a complex new binary format and adding it to the TL. Those days should be behind us, although we need to support legacy formats in some cases.

Since you both want to mention HOB, that's fine. But please add the FDT link back.

Reply:
Here's a hexdump of a random FDT, let's see how opaque it is:
```
00000000: d00d feed 0001 69a0 0000 0038 0001 5a84  ......i....8..Z.
00000010: 0000 0028 0000 0011 0000 0010 0000 0000  ...(............
00000020: 0000 0f1c 0001 5a4c 0000 0000 0000 0000  ......ZL........
[...]
00001080: 0000 0003 0000 0008 0000 01c5 0000 000a  ................
00001090: 0000 0000 0000 0003 0000 0004 0000 00f9  ................
000010a0: 0000 001a 0000 0001 6c32 2d63 6163 6865  ........l2-cache
000010b0: 0000 0000 0000 0003 0000 0006 0000 0032  ...............2
000010c0: 6361 6368 6500 0000 0000 0003 0000 0004  cache...........
000010d0: 0000 01a5 0000 000b 0000 0003 0000 0004  ................
000010e0: 0000 00f9 0000 000f 0000 0002 0000 0002  ................
000010f0: 0000 0001 6370 7540 3530 3000 0000 0003  ....cpu@500.....
00001100: 0000 0004 0000 012c 6370 7500 0000 0003  .......,cpu.....
[...]
00015eb0: 6c2d 6e61 6d65 7300 7069 6e63 7472 6c2d  l-names.pinctrl-
00015ec0: 3000 7069 6e63 7472 6c2d 3100 766d 6d63  0.pinctrl-1.vmmc
00015ed0: 2d73 7570 706c 7900 7671 6d6d 632d 7375  -supply.vqmmc-su
00015ee0: 7070 6c79 0072 6571 7569 7265 642d 6f70  pply.required-op
00015ef0: 7073 006f 7070 2d61 7667 2d6b 4270 7300  ps.opp-avg-kBps.
00015f00: 6373 2d67 7069 6f73 0073 7069 2d6d 6178  cs-gpios.spi-max
00015f10: 2d66 7265 7175 656e 6379 0067 7069 6f2d  -frequency.gpio-
[...]
```
Certainly completely different from a binary blob!

Updated to add links for both examples.

Reply:
You could try fdtdump ?

Reply:
Yes, and I can write dump tools for any other imaginable data structure too. I still fail to see what inherently makes the FDT "self-describing" here.

Reply:
You can see what fields are present and read the values. With a binary blob you have no idea what the data means.

You can't write a dump tool if you don't know the format of the data. FDT forces that into the open, at least to an extent.

It is not fully self-describing, since the schema is not included. But even without that, it helps a lot.